### ISSUE TITLE
Bug: no-unnecessary-type-assertion undefined union false positives

### ISSUE THREAD

**Issue by dl-pt** [typescript-eslint/typescript-eslint#12249]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=6.0.2&fileType=.tsx&code=MYewdgzgLgBKZQKYA8oC4YG8YFtFQEMMAlRUAJwBMAeacgSzAHMAaGAVzAGswQB3MAD4YAHw5hKiAGaNElGAF8YAXiy58RLEoUBuALAAoeNHWEVccElQA6PGYIQ1EYAAtEOAgH4MIAEYArMlglMU5JGTA5fQMgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYkACAXAngA4oDG0AlobgLQrzkB2uA9AwPbWwMOIkrIBDaPmoFi1AcmSJoucmwbooM6G2iQAviA1A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
const context: { meta: Record<string, unknown> | undefined } = { meta: {} };
const meta = context.meta as { schema?: object } | undefined;
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-type-assertion": "error"
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

No error since the type assertion is necessary and makes sense.

### Actual Result

Linter error and the auto-fix breaks the code (provokes a tsc error).

**Comment by bradzacher** [comment]
Extra info to show it is a bug with unions specifically - if you remove the `| undefined` from both the property type and the assertion then the rule does not error.

```ts
declare const context: { meta: Record<string, unknown> };
const meta = context.meta as { schema?: object };
```
[playground](https://typescript-eslint.io/play/#ts=6.0.2&fileType=.tsx&code=CYUwxgNghgTiAEYD2A7AzgF0ajIAeGAXPAN7wC2IGUxASuEjMADyYwCWKA5gDTwCuKANYokAdxQA%2BeAF8A3AFgAUMnRZK1eAF5sKXAQB0GqPChpS8NGAAWIclAD8xJACMAVuCzzlAeh-wAwKDg4IA9B2UgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYkACAXAngA4oDG0AlobgLQrzkB2uA9AwPbWwMOIkrIBDaPmoFi1AcmSJoucmwbooM6G2iQAviA1A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)
